File: 2618-maximize-the-minimum-powered-city/maximize-the-minimum-powered-city.go
func maxPower(stations []int, r int, k int) int64 {
	n := len(stations)
	power := make([]int64, n)
	for i := 0; i < len(stations); i++ {
		left, right := i-r, i+r+1
		if left < 0 {
			left = 0
		}
		power[left] += int64(stations[i])
		if right < n {
			power[right] -= int64(stations[i])
		}
	}

	minP, maxP := power[0], power[0]
	for i := 1; i < n; i++ {
		power[i] += power[i-1]
		minP = min(minP, power[i])
		maxP = max(maxP, power[i])
	}

	left, right := minP, maxP+int64(k)
	var res int64
	for left <= right {
		mid := left + (right-left)/2
		if check(power, mid, r, int64(k)) {
			res = mid
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
    
	return res
}

func check(power []int64, target int64, r int, k int64) bool {
	q := list.New() 
	var extra int64
	for i := 0; i < len(power); i++ {
		for q.Len() > 0 && q.Front().Value.([2]int64)[0] <= int64(i) {
			extra += q.Remove(q.Front()).([2]int64)[1]
		}
		if power[i]+extra < target {
			needed := target - power[i] - extra
			if k < needed {
				return false
			}
			extra += needed
			k -= needed
			if i+r+1 < len(power) {
				q.PushBack([2]int64{int64(i + 2*r + 1), -needed})
			}
		}
	}
	return true
}

func max(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

func min(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}

File: 2618-maximize-the-minimum-powered-city/maximize-the-minimum-powered-city.py
class Solution:
    def maxPower(self, stations: List[int], r: int, k: int) -> int:
        n = len(stations)
        left, right = 0, k + sum(stations)
        while left <= right:
            x = (left + right) // 2
            use = 0
            v = stations.copy()
            s = sum(stations[0:r])
            for i in range(n):
                t = n - 1 if n - 1 < i + r else i + r
                if i + r < n:
                    s += v[i + r]
                if i - r > 0:
                    s -= v[i - r - 1]
                diff = x - s if x - s > 0 else 0
                v[t] += diff
                s += diff
                use += diff
            if use <= k:
                left = x + 1
            else:
                right = x - 1
        return right
